core/src/main/scala/kafka/coordinator/GroupCoordinator.scala

```diff
@@ -0,0 +1,588 @@
+          if (group.is(Dead)) {
+            partitions.map {case topicAndPartition => (topicAndPartition, OffsetMetadataAndError.UnknownMember)}.toMap
+          } else {
+            offsetManager.getOffsets(groupId, partitions)
+          }
+        }
+      }
+    }
+  }
+
+  def handleGroupImmigration(offsetTopicPartitionId: Int) = {
+    // TODO we may need to add more logic in KAFKA-2017
+    offsetManager.loadOffsetsFromLog(offsetTopicPartitionId)
+  }
+
+  def handleGroupEmigration(offsetTopicPartitionId: Int) = {
+    // TODO we may need to add more logic in KAFKA-2017
+    offsetManager.removeOffsetsFromCacheForPartition(offsetTopicPartitionId)
+  }
+
+  private def propagateAssignment(group: GroupMetadata,
+                                  assignment: Map[String, Array[Byte]]) {
+    for (member <- group.allMembers) {
+      member.assignment = assignment.getOrElse(member.memberId, Array.empty[Byte])
+      if (member.awaitingSyncCallback != null) {
+        member.awaitingSyncCallback(member.assignment, Errors.NONE.code)
+        member.awaitingSyncCallback = null
+      }
+    }
+  }
+
+  /**
+   * Complete existing DelayedHeartbeats for the given member and schedule the next one
+   */
+  private def completeAndScheduleNextHeartbeatExpiration(group: GroupMetadata, member: MemberMetadata) {
+    // complete current heartbeat expectation
+    member.latestHeartbeat = SystemTime.milliseconds
+    val memberKey = ConsumerKey(member.groupId, member.memberId)
+    heartbeatPurgatory.checkAndComplete(memberKey)
+
+    // reschedule the next heartbeat expiration deadline
+    val newHeartbeatDeadline = member.latestHeartbeat + member.sessionTimeoutMs
+    val delayedHeartbeat = new DelayedHeartbeat(this, group, member, newHeartbeatDeadline, member.sessionTimeoutMs)
+    heartbeatPurgatory.tryCompleteElseWatch(delayedHeartbeat, Seq(memberKey))
+  }
+
+  private def addMember(memberId: String,
+                        sessionTimeoutMs: Int,
+                        metadata: Array[Byte],
+                        subProtocols: Set[String],
+                        group: GroupMetadata) = {
+    val member = new MemberMetadata(memberId, group.groupId, sessionTimeoutMs, metadata, subProtocols)
+    group.add(member.memberId, member)
+    member
+  }
+
+  private def removeMember(group: GroupMetadata, member: MemberMetadata) {
+    group.remove(member.memberId)
+  }
+
+  private def updateMember(group: GroupMetadata, member: MemberMetadata, metadata: Array[Byte], subProtocols: Set[String]) {
+    group.remove(member.memberId)
+    member.metadata = metadata
+    member.subProtocols = subProtocols
+    group.add(member.memberId, member)
+  }
+
+  private def maybePrepareRebalance(group: GroupMetadata) {
+    group synchronized {
+      if (group.canRebalance)
+        prepareRebalance(group)
+    }
+  }
+
+  private def prepareRebalance(group: GroupMetadata) {
+    // if any members are awaiting sync, cancel their request and have them rejoin
```

Comment:
Since maybePrepareRebalance and proporgateAssignment are all synchronized on the group, when the group is still in awaiting sync either all members still have their callback yet to trigger, or all of them should have sent the response in callback right?

Also the comments are not very accurate: if the group is in the awaiting sync, cancel the sync response for all of them if possible to have all its members rejoin.